controllers/secret_controller.go

```diff
@@ -162,36 +163,58 @@ func (r *SecretReconciler) Reconcile(ctx context.Context, request ctrl.Request)
+		// userdata secret data does not match what is expected, implies private key has changed
+		// Modify annotations on nodes configured with the previous private key
 		nodes := &core.NodeList{}
 		err = r.client.List(ctx, nodes, client.MatchingLabels{core.LabelOSStable: "windows"})
 		if err != nil {
 			return reconcile.Result{}, errors.Wrapf(err, "error getting node list")
 		}
+
 		expectedPubKeyAnno := nodeconfig.CreatePubKeyHashAnnotation(keySigner.PublicKey())
-		patchData, err := annotations.GenerateAddPatch(map[string]string{nodeconfig.PubKeyHashAnnotation: ""})
+		privateKeyBytes, err := secrets.GetPrivateKey(kubeTypes.NamespacedName{Namespace: r.watchNamespace,
+			Name: secrets.PrivateKeySecret}, r.client)
 		if err != nil {
-			return reconcile.Result{}, errors.Wrapf(err, "error creating public key annotation add request")
+			return reconcile.Result{}, err
 		}
 		for _, node := range nodes.Items {
-			// Only clear the annotation for Nodes associated with Machines, as the clearing of the annotation is used
-			// solely to kick off the deletion and recreation of Machines.
+			annotationsToApply := make(map[string]string)
 			if _, present := node.Annotations[BYOHAnnotation]; present {
-				continue
-			}
+				// For BYOH nodes, update the username annotation and public key hash annotation using new private key
+				instanceUsername, err := r.usernameFromNode(ctx, &node, nodes)
+				if err != nil {
+					return reconcile.Result{}, err
+				}
+				usernameCipher, err := secrets.Encrypt(instanceUsername, privateKeyBytes)
+				if err != nil {
+					return reconcile.Result{}, err
+				}
+				// Obfuscate newline characters to make encrypted data compatible as JSON Patch request body
+				patchCombatibleCipher := strings.Replace(usernameCipher, "\n", "<newline>", -1)
 
-			existingPubKeyAnno := node.Annotations[nodeconfig.PubKeyHashAnnotation]
-			if existingPubKeyAnno == expectedPubKeyAnno {
-				continue
-			}
-			node.Annotations[nodeconfig.PubKeyHashAnnotation] = ""
-			err = r.client.Patch(ctx, &node, client.RawPatch(kubeTypes.JSONPatchType, patchData))
-			if err != nil {
-				return reconcile.Result{}, errors.Wrapf(err, "error clearing public key annotation on node %s",
-					node.GetName())
+				annotationsToApply = map[string]string{
+					UsernameAnnotation:              patchCombatibleCipher,
+					nodeconfig.PubKeyHashAnnotation: expectedPubKeyAnno,
+				}
+				if err := r.applyAnnotations(ctx, node, annotationsToApply); err != nil {
+					return reconcile.Result{},
+						errors.Wrapf(err, "error updating username & public key annotations on node %s", node.GetName())
+				}
+			} else {
+				// For Nodes associated with Machines, clear the public key annotation, as the clearing of the
+				// annotation is used solely to kick off the deletion and recreation of Machines.
+				existingPubKeyAnno := node.Annotations[nodeconfig.PubKeyHashAnnotation]
+				if existingPubKeyAnno == expectedPubKeyAnno {
```

Comment:
Can't you do a similar check for the encrypted username?

Reply:
Revisiting this, I don't think that this check on the username is necessary -- checking the public key annotation is sufficient for both. If the public key anno changed, that means the private key secret has changed, so both `PubKeyHashAnnotation` and `UsernameAnnotation` must be updated.

Reply:
What if the user changed the private key secret but didn't actually update the key? Like added a label or annotation.

Reply:
Both annotations' values are made using the bytes present in private key secret data. So no change necessary if a label/annotation is added